### PR TITLE
Support Envoy buffer filter

### PR DESCRIPTION
**Pull request by Sodman**
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/2444

### CODE REVIEW

File: projects/gloo/api/v1/options.proto [hunk 1]
```diff
@@ -107,6 +108,13 @@ message HttpListenerOptions {
 
     // Enterprise-only: Proxy latency
     envoy.config.filter.http.proxylatency.v2.ProxyLatency proxy_latency = 9;
+
+    // Buffer can be used to set the maximum request size
+    // that the filter will buffer before the connection
+    // manager will stop buffering and return a 413 response.
+    envoy.config.filter.http.buffer.v2.Buffer buffer = 12;
+
+    envoy.config.filter.http.buffer.v2.BufferPerRoute buffer_per_route = 13;
```
Comment (kdorosh) [inline projects/gloo/api/v1/options.proto:117]: the `BufferPerRoute` should live on `RouteOptions` message

**Comment by Sodman** [comment]
Tests are all passing and the config looks to be generated correctly, but I'd like to manually test this in my local environment a bit more thoroughly before merging.

File: changelog/v1.4.0-beta14/support-buffer-filter.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+changelog:
+    - type: NEW_FEATURE
+      issueLink: https://github.com/solo-io/gloo/issues/2444
+      description: >
+        Gloo now supports enabling the Envoy buffer filter by configuring `spec.httpGateway.options.buffer` of the desired Gateway.
```
Comment (kdorosh) [inline changelog/v1.4.0-beta14/support-buffer-filter.yaml:5]: also supports on virtual services at virtual host or route level, and on weighted destinations

File: projects/gloo/api/external/envoy/config/filter/http/buffer/v2/buffer.proto [hunk 3]
```diff
@@ -0,0 +1,48 @@
+// copied from https://raw.githubusercontent.com/envoyproxy/envoy/549164c42cae84b59154ca4c36009e408aa10b52/api/envoy/config/filter/http/buffer/v2/buffer.proto
+
+syntax = "proto3";
+
+package envoy.config.filter.http.buffer.v2;
```
Comment (kdorosh) [inline projects/gloo/api/external/envoy/config/filter/http/buffer/v2/buffer.proto:5]: we're actively moving to v3 envoy API, let's move to v3 in this PR as well https://github.com/envoyproxy/envoy/blob/3b52fc36373272902d9817f0db97dd2fccc40784/api/envoy/extensions/filters/http/buffer/v3/buffer.proto#L3

File: projects/gloo/pkg/plugins/buffer/plugin.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	v1 "github.com/solo-io/gloo/projects/gloo/pkg/api/v1"
+	"github.com/solo-io/gloo/projects/gloo/pkg/plugins"
+	"github.com/solo-io/gloo/projects/gloo/pkg/plugins/pluginutils"
+	"github.com/solo-io/solo-kit/pkg/api/v1/control-plane/util"
+)
+
+// filter should be called after routing decision has been made
+var pluginStage = plugins.DuringStage(plugins.RouteStage)
+
+const FilterName = util.Buffer
```
Comment (kdorosh) [inline projects/gloo/pkg/plugins/buffer/plugin.go:20]: let's update this to the newer, non-deprecated filter name `envoy.filters.http.buffer`

File: projects/gloo/pkg/plugins/buffer/plugin_test.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+		}, out)
+
+		var cfg envoybuffer.BufferPerRoute
+		err = conversion.StructToMessage(out.GetPerFilterConfig()[FilterName], &cfg)
+
+		Expect(err).NotTo(HaveOccurred())
+		Expect(cfg.GetBuffer().GetMaxRequestBytes().GetValue()).To(Equal(uint32(4098)))
+	})
+
+	It("allows vhost specific buffer config", func() {
```
Comment (kdorosh) [inline projects/gloo/pkg/plugins/buffer/plugin_test.go:134]: these two tests need their names swapped

File: projects/gloo/pkg/plugins/buffer/plugin_test.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+		}, out)
+
+		var cfg envoybuffer.BufferPerRoute
+		err = conversion.StructToMessage(out.GetPerFilterConfig()[FilterName], &cfg)
+
+		Expect(err).NotTo(HaveOccurred())
+		Expect(cfg.GetDisabled()).To(Equal(true))
+	})
+
+	It("allows weighted destination specific disabling of buffer", func() {
```
Comment (kdorosh) [inline projects/gloo/pkg/plugins/buffer/plugin_test.go:174]: switch these test names as well

**Comment by kdorosh** [comment]
/kick possible flake in codegen here https://github.com/solo-io/gloo/pull/3086/checks?check_run_id=731505738 and in regression tests here https://console.cloud.google.com/cloud-build/builds/933a7e88-9e6d-4f4d-a5a2-1792958a9104;step=14?project=solo-public

**Comment by kdorosh** [comment]
/kick codegen failed on teardown with no logs here https://github.com/solo-io/gloo/pull/3086/checks?check_run_id=732174219 need to figure out why

also CI failed with probably flake here https://console.cloud.google.com/cloud-build/builds/14050468-4c1c-4982-9f04-9e3068773ccd;step=14?project=solo-public

**Comment by Sodman** [comment]
/kick codegen flake - https://github.com/solo-io/gloo/pull/3086/checks?check_run_id=734980431

**Pull request merged**